native-assets/NFT/project/contracts/src/main.sw

```diff
@@ -0,0 +1,641 @@
+    let src3_abi = abi(SRC3, CONTRACT_ID);
+    let src20_abi = abi(SRC20, CONTRACT_ID);
+
+    let recipient = Identity::ContractId(ContractId::from(CONTRACT_ID));
+    let sub_id1 = 0x0000000000000000000000000000000000000000000000000000000000000001;
+    let sub_id2 = 0x0000000000000000000000000000000000000000000000000000000000000002;
+
+    assert(src20_abi.total_assets() == 0);
+
+    src3_abi.mint(recipient, sub_id1, 1);
+    assert(src20_abi.total_assets() == 1);
+
+    src3_abi.mint(recipient, sub_id2, 1);
+    assert(src20_abi.total_assets() == 2);
+}
+
+#[test]
+fn test_total_supply() {
+    use std::constants::ZERO_B256;
+
+    let src3_abi = abi(SRC3, CONTRACT_ID);
+    let src20_abi = abi(SRC20, CONTRACT_ID);
+
+    let recipient = Identity::ContractId(ContractId::from(CONTRACT_ID));
+    let sub_id = ZERO_B256;
+    let asset_id = AssetId::new(ContractId::from(CONTRACT_ID), sub_id);
+
+    assert(src20_abi.total_supply(asset_id).is_none());
+
+    src3_abi.mint(recipient, sub_id, 1);
+    assert(src20_abi.total_supply(asset_id).unwrap() == 1);
+}
+
+#[test]
+fn test_name() {
+    use std::constants::ZERO_B256;
+    
+    let src20_abi = abi(SRC20, CONTRACT_ID);
+    let attributes_abi = abi(SetTokenAttributes, CONTRACT_ID);
+
+    let sub_id = ZERO_B256;
+    let asset_id = AssetId::new(ContractId::from(CONTRACT_ID), sub_id);
+    let name = String::from_ascii_str("Fuel Token");
+
+    assert(src20_abi.name(asset_id).is_none());
+
+    attributes_abi.set_name(asset_id, name);
+    assert(src20_abi.name(asset_id).unwrap().as_bytes() == name.as_bytes());
+}
+
+#[test(should_revert)]
+fn test_revert_set_name_twice() {
+    use std::constants::ZERO_B256;
+    
+    let attributes_abi = abi(SetTokenAttributes, CONTRACT_ID);
+    let sub_id = ZERO_B256;
+    let asset_id = AssetId::new(ContractId::from(CONTRACT_ID), sub_id);
+    let name = String::from_ascii_str("Fuel Token");
+
+    attributes_abi.set_name(asset_id, name);
+    attributes_abi.set_name(asset_id, name);
+}
+
+#[test]
+fn test_symbol() {
+    use std::constants::ZERO_B256;
+
+    let src20_abi = abi(SRC20, CONTRACT_ID);
+    let attributes_abi = abi(SetTokenAttributes, CONTRACT_ID);
+
+    let sub_id = ZERO_B256;
+    let asset_id = AssetId::new(ContractId::from(CONTRACT_ID), sub_id);
+    let symbol = String::from_ascii_str("FUEL");
+
+    assert(src20_abi.symbol(asset_id).is_none());
+
+    attributes_abi.set_symbol(asset_id, symbol);
+    assert(src20_abi.symbol(asset_id).unwrap().as_bytes() == symbol.as_bytes());
+}
+
+#[test(should_revert)]
+fn test_revert_set_symbol_twice() {
+    use std::constants::ZERO_B256;
+
+    let attributes_abi = abi(SetTokenAttributes, CONTRACT_ID);
+    let sub_id = ZERO_B256;
+    let asset_id = AssetId::new(ContractId::from(CONTRACT_ID), sub_id);
+    let symbol = String::from_ascii_str("FUEL");
+
+    attributes_abi.set_symbol(asset_id, symbol);
+    attributes_abi.set_symbol(asset_id, symbol);
+}
+
+#[test]
+fn test_decimals() {
```

Comment:
Should panic test?

Reply:
No, this tests the getter. It's the setter that always panics

Reply:
I meant that we should add a test for the panic case